lib/repo/repository.go

```diff
@@ -413,3 +429,142 @@ func VirtualBoxRepositoryEdison() Repository {
 	//return NewGenericRepository("https://s3-ap-northeast-1.amazonaws.com/isaax-distro/vm/sd/0.1.0/isaax-box-sd.zip", "0.0.1", "virtualbox/sd/")
 	return NewGenericRepository(rp.GetURL(), rp.GetVersion(), rp.Dir())
 }
+
+// DownloadNewVersion downloads the latest version based on the current release and skips this step if up to date
+func DownloadNewVersion(name, version, dst string) (string, error) {
+	zipMethod := "zip"
+	if runtime.GOOS == "linux" {
+		zipMethod = "tar.gz"
+	}
+
+	fileName := fmt.Sprintf("%s_%s_%s_%s", name, version, runtime.GOOS, runtime.GOARCH)
+
+	_, version, err := GetIoTItVersionMD5(runtime.GOOS, runtime.GOARCH, version)
+	if err != nil || version == "" {
+		return "", err
+	}
+
+	url := fmt.Sprintf("https://cdn.isaax.io/%s/%s/%s/%s.%s", name, CurrentRelease(version), runtime.GOOS, fileName, zipMethod)
+
+	wg := &sync.WaitGroup{}
+	imgName, bar, err := help.DownloadFromUrlWithAttemptsAsync(url, dst, 5, wg)
+	if err != nil {
+		return fileName, err
+	}
+	bar.Prefix(fmt.Sprintf("[+] Download %-15s", imgName))
+	bar.Start()
+	wg.Wait()
+	bar.Finish()
+	time.Sleep(time.Second)
+
+	fmt.Println("[+] Extracting into ", dst)
+	if runtime.GOOS == "linux" {
+		if err := exec.Command("tar", "xvf", dst+help.Separator()+fileName, "-C", dst).Run(); err != nil {
+			fmt.Println("[-] ", err)
+			return fileName, err
+		}
+	}
+
+	if err := exec.Command("unzip", "-o", dst+help.Separator()+fileName, "-d", dst).Run(); err != nil {
+		fmt.Println("[-] ", err)
+		return fileName, err
+	}
+
+	return fileName, nil
+}
+
+// CurrentRelease detects whether release is stable or latest
+func CurrentRelease(version string) (release string) {
+	r := Latest
+	match, _ := regexp.Compile(`^[\d|_]+\.[\d|_]+\.[\d|_]+$`)
+	if match.MatchString(version) {
+		r = Stable
+	}
+
+	return r
+}
+
+// GetVersionLexem parses string lexems into comparable parts
+func GetVersionLexem(token string, seps ...string) []string {
+	var lexs []string
+	for i, sep := range seps {
+		if i == 0 {
+			lexs = strings.Split(token, sep)
+			continue
+		}
+
+		var tmp []string
+		for _, lex := range lexs {
+			tmp = append(tmp, strings.Split(lex, sep)...)
+		}
+
+		lexs = tmp
+	}
+
+	return lexs
+}
+
+func IsVersionUpToDate(v1, v2 string) (bool, error) {
+	vlex1 := GetVersionLexem(v1, ".", "_", "-")
+	vlex2 := GetVersionLexem(v2, ".", "_", "-")
+
+	for i := 0; i < len(vlex1) && i < len(vlex2); i++ {
+		n1, err := strconv.Atoi(vlex1[i])
+		if err != nil {
+			return false, err
+		}
+
+		n2, err := strconv.Atoi(vlex2[i])
+		if err != nil {
+			return false, err
+		}
+
+		if n1 == n2 {
+			continue
+		}
+
+		return n1 > n2, nil
+	}
+
+	// not reachable
+	return false, nil
+}
+
+func GetIoTItVersionMD5(oss, arch, version string) (hash string, repoVersion string, err error) {
```

Comment:
__[golint]__
* [golint] exported function GetIoTItVersionMD5 should have comment or be unexported

<!-- Comment by [SideCI](https://sideci.com) -->